### ISSUE TITLE
_writev doesn't wait for chunk log callbacks before calling its own callback

### ISSUE THREAD

**Issue by laganojunior** [winstonjs/winston-transport#49]
Function in question: https://github.com/winstonjs/winston-transport/blob/46db8f3c8cd8b106ade8d7e04a191ee388683d60/index.js#L117-L166

This implementation of _writev calls `callback` before all the callbacks of `this.log(chunks[i]...)` are called. This means that the transport stream may seem finished even if the underlying implementation hasn't finished writing out all the chunks yet.

We ran into this when we saw that during tests that when we close a winston logger, we would sometimes see a `finish` event be emitted from the logger before our transport finished writing its files.

We're currently working around this by implementing our own logv for our transport since that gives us control over calling the callback to `_writev` only when each chunk is finished.

I'm happy to try to construct a minimal example and/or provide a merge request to fix this. I just want to check here if this is something that's done by design or otherwise known.

**Comment by laganojunior** [comment]
This may be related to other people reporting similar observations of `'finished'` events being emitted early?

- https://github.com/winstonjs/winston/issues/1250#issuecomment-450658764
- https://github.com/winstonjs/winston/issues/1504

**Comment by laganojunior** [comment]
Also this too:

https://github.com/winstonjs/winston/issues/228#issuecomment-511731825

I imagine this workaround works because they're bypassing the `finished` event of `TransportStream` and instead using the `finished` event of the underlying file write stream which probably does wait for file writes before emitting `finished. Unfortunately, this workaround probably doesn't work for generic custom transports unless they happen to define an internal `_stream_ variable.

**Comment by laganojunior** [comment]
Made a commit to add a test to demonstrate the issue and a patch to fix it.

**Comment by danielweck** [comment]
Yes, you are correct, my workaround ; which is itself inspired by other implementations reported in the issue tracker ; relies on the file Transport allowing access to the "private" `_stream` event emitter:

`transport._stream.once('finish', finishedCallback); transport._stream.end();`

(instead of `transport.once('finish', finishedCallback); transport.close();`, which currently isn't a reliable mechanism for a clean detection of Winston completion)

The worst part of this hack is the necessity to run the workaround for each registered transport, but as I could not find a reliable Winston accessor I had to allocate a `_transports` variable (`for (const transport of _transports) { ... }`).

I am sure there is a better way to implement this, but whatever worked at the time did the job for me. I simply had no time to look into Winston's internals.

https://github.com/winstonjs/winston/issues/228#issuecomment-511731825

**Comment by derekngo** [comment]
+1
